plugins/chat/assets/javascripts/discourse/components/chat-search.gjs

```diff
@@ -0,0 +1,291 @@
+import ChatChannel from "discourse/plugins/chat/discourse/models/chat-channel";
+import ChatMessage from "discourse/plugins/chat/discourse/models/chat-message";
+
+export default class ChatSearch extends Component {
+  @service router;
+
+  @tracked currentSort = "relevance";
+  @tracked messages = [];
+  @tracked isLoading = false;
+  @tracked offset = 0;
+  @tracked hasMoreResults = false;
+  @tracked lastQuery = null;
+
+  limit = 20;
+
+  @bind
+  async searchMessages(resetResults = true) {
+    if (isBlank(this.args.query)) {
+      this.messages = [];
+      this.offset = 0;
+      this.hasMoreResults = false;
+      return;
+    }
+
+    if (this.isLoading) {
+      return;
+    }
+
+    this.isLoading = true;
+
+    if (resetResults) {
+      this.offset = 0;
+      this.messages = [];
+    }
+
+    try {
+      const response = await ajax("/chat/api/search", {
+        data: {
+          query: this.args.query,
+          channel_id: this.args.scopedChannelId,
+          sort: this.currentSort,
+          offset: this.offset,
+          limit: this.limit,
+        },
+      });
+
+      const newMessages = (response.messages || []).map((messageObject) => {
+        return ChatMessage.create(
+          ChatChannel.create(messageObject.channel),
+          messageObject
+        );
+      });
+
+      if (resetResults) {
+        this.messages = newMessages;
+      } else {
+        this.messages = [...this.messages, ...newMessages];
+      }
+
+      this.offset += newMessages.length;
+      this.hasMoreResults = response.meta?.has_more || false;
+
+      return this.messages;
+    } finally {
+      this.isLoading = false;
+    }
+  }
+
+  @action
+  sortLabel(sort) {
+    return i18n(`chat.search.sort.${sort}`);
+  }
+
+  @action
+  setCurrentSorting(sort, closeMenu) {
+    this.currentSort = sort;
+    closeMenu();
+    this.searchMessages(true);
+  }
+
+  @action
+  onFilterChange(element) {
+    const value = element.target.value;
+    this.debouncedHandler = discourseDebounce(
+      this.debounceFilterChange,
+      value === "" ? null : value,
+      INPUT_DELAY
+    );
+  }
+
+  @action
+  async loadMore() {
+    if (!this.hasMoreResults || this.isLoading) {
+      return;
+    }
+
+    await this.searchMessages(false);
+  }
+
+  @action
+  debounceFilterChange(query) {
+    this.router.replaceWith({ queryParams: { q: query } });
+  }
+
+  @action
+  loadExistingQuery() {
+    this.checkQueryChange();
+  }
+
+  @action
+  checkQueryChange() {
+    if (this.lastQuery !== this.args.query) {
+      this.lastQuery = this.args.query;
+      this.searchMessages(true);
+    }
+  }
+
+  @action
+  visitMessage(message) {
+    let url;
+
+    if (message.threadId) {
+      url = getURL(
+        `/chat/c/-/${message.channel.id}/t/${message.threadId}/${message.id}`
+      );
+    } else {
+      url = getURL(`/chat/c/-/${message.channel.id}/${message.id}`);
+    }
```

Comment:
IMO we could add a `url()` getter to the `ChatMessage` model for this, then you could shorten this function to:

```js
visitMessage(message) {
  DiscourseURL.routeTo(message.url);
}
```